proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,222 @@
+epoch boundary. When reaching block height `N` after the start of the
+`reward calculation phase`, the bank will mark it the end the `reward calculation phase`.
+A conservative value of 1,000 is chosen for `N` to accommodate the most of the
+validators on the network.
+
+At this point, the reward computation result should be available. In case that
+the reward computation result is not available yet at the `N_th_` block-height,
+the bank runtime will block and wait until the computation is finished, before
+processing any blocks with height greater than `N`.
+
+At the end of the reward computation phase (end of block height `N`), a set of
+sysvar accounts will be updated - `EpochRewardProof` and `RewardReserve`.
+
+`EpochRewardProof` sysvar account maintains a history of epoch rewards. The
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block at block height `height` in `Reward Distribution`
+phase. The set of stake rewards paid from `EpochRewardsReserve`
+sysvar accounts form a complete partition of the total stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   height: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards (in lamports) to be distributed in this block.
+`height` is the reward distribution absolute block height.
+`hash` is the hash of all distributed rewards for this block.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick without much performance impact for the block time. Given
+that there are 64 ticks per block, each block will store 4,096 accounts.
+The total number of blocks to distributed rewards is
+
+```
+M = ceil(num_stake_accounts/4096)
+```
+
+For each block, `i` in reward distribution phase, the rewards correspond to
+`EpochRewardReserve` sysvar account with `height=epoch_height+N+i` will be paid
+out.
+
```

Comment:
The `RewardHistoryEntry::distributed` field gets updated as well to account for the new distribution right?